src/algebra/lie/engel.lean

```diff
@@ -0,0 +1,279 @@
+  { simp, },
+  { exact (l.succ_mul n).symm ▸ lcs_le_lcs_of_is_nilpotent_span_sup_eq_top hxI hn ih, },
+end
+
+end lie_submodule
+
+section lie_algebra
+
+open lie_module (hiding is_nilpotent)
+
+variables (R L)
+
+/-- A Lie algebra `L` is said to be Engelian if a sufficient condition for any `L`-module `M` to
+be nilpotent is that image of the map `L → End(M)` consists of nilpotent elements.
+
+Engel's theorem `lie_algebra.is_engelian_of_is_noetherian` states that any Noetherian Lie algebra is
+Engelian. -/
+def lie_algebra.is_engelian : Prop :=
+  ∀ (M : Type u₄) [add_comm_group M], by exactI ∀ [module R M] [lie_ring_module L M], by exactI ∀
+    [lie_module R L M], by exactI ∀ (h : ∀ (x : L), is_nilpotent (to_endomorphism R L M x)),
+    lie_module.is_nilpotent R L M
+
+variables {R L}
+
+lemma lie_algebra.is_engelian_of_subsingleton [subsingleton L] : lie_algebra.is_engelian R L :=
+begin
+  intros M _i1 _i2 _i3 _i4 h,
+  use 1,
+  suffices : (⊤ : lie_ideal R L) = ⊥, { simp [this], },
+  haveI := (lie_submodule.subsingleton_iff R L L).mpr (by apply_instance),
+  apply subsingleton.elim,
+end
+
+lemma function.surjective.is_engelian
+  {f : L →ₗ⁅R⁆ L₂} (hf : function.surjective f) (h : lie_algebra.is_engelian.{u₁ u₂ u₄} R L) :
+  lie_algebra.is_engelian.{u₁ u₃ u₄} R L₂ :=
+begin
+  introsI M _i1 _i2 _i3 _i4 h',
+  letI : lie_ring_module L M := lie_ring_module.comp_lie_hom M f,
+  letI : lie_module R L M := comp_lie_hom M f,
+  have hnp : ∀ x, is_nilpotent (to_endomorphism R L M x) := λ x, h' (f x),
+  have surj_id : function.surjective (linear_map.id : M →ₗ[R] M) := function.surjective_id,
+  haveI : lie_module.is_nilpotent R L M := h M hnp,
+  apply hf.lie_module_is_nilpotent surj_id,
+  simp,
+end
+
+lemma lie_equiv.lie_algebra_is_engelian_iff (e : L ≃ₗ⁅R⁆ L₂) :
+  lie_algebra.is_engelian.{u₁ u₂ u₄} R L ↔ lie_algebra.is_engelian.{u₁ u₃ u₄} R L₂ :=
+⟨e.surjective.is_engelian, e.symm.surjective.is_engelian⟩
+
+lemma lie_algebra.exists_engelian_lie_subalgebra_of_lt_normalizer
+  {K : lie_subalgebra R L} (hK₁ : lie_algebra.is_engelian.{u₁ u₂ u₄} R K) (hK₂ : K < K.normalizer) :
+  ∃ (K' : lie_subalgebra R L) (hK' : lie_algebra.is_engelian.{u₁ u₂ u₄} R K'), K < K' :=
+begin
+  obtain ⟨x, hx₁, hx₂⟩ := set_like.exists_of_lt hK₂,
+  let K' : lie_subalgebra R L :=
+  { lie_mem' := λ y z, lie_subalgebra.lie_mem_sup_of_mem_normalizer hx₁,
+    .. (R ∙ x) ⊔ (K : submodule R L) },
+  have hxK' : x ∈ K' := submodule.mem_sup_left (submodule.subset_span (set.mem_singleton _)),
+  have hKK' : K ≤ K' := (lie_subalgebra.coe_submodule_le_coe_submodule K K').mp le_sup_right,
+  have hK' : K' ≤ K.normalizer,
+  { rw ← lie_subalgebra.coe_submodule_le_coe_submodule,
+    exact sup_le ((submodule.span_singleton_le_iff_mem _ _).mpr hx₁) (le_of_lt hK₂), },
+  refine ⟨K', _, lt_iff_le_and_ne.mpr ⟨hKK', λ contra, hx₂ (contra.symm ▸ hxK')⟩⟩,
+  introsI M _i1 _i2 _i3 _i4 h,
+  obtain ⟨I, hI₁⟩ := lie_subalgebra.exists_nested_lie_ideal_of_le_normalizer hKK' hK',
```

Comment:
`obtain` supports giving explicit type ascriptions to the arguments. I've been able to read this proof without VScode up to this point. Could you please write the type of `hI₁` explicitly in the code? I think that would really help.

Reply:
Sure, done!